rfcs/0000-ckb-softfork-activation/0000-ckb-softfork-activation.md

```diff
@@ -0,0 +1,132 @@
+5. The `threshold` specifies the minimum ratio of block per epoch, which indicate the locked_in of the softfork during the epoch.
+6. The `minimum_activation_epoch` specifies the epoch at which the softfork is allowed to become active.
+
+### Selection guidelines
+The following guidelines are suggested for selecting these parameters for a softfork:
+
+1. `name` should be selected such that no two softforks, concurrent or otherwise, ever use the same name.
+2. `bit` should be selected such that no two concurrent softforks use the same bit.
+3. `start_epoch` can be set soon after software with parameters is expected to be released.
+4. `timeout_epoch` should be set to an epoch when it is considered reasonable to expect the entire economy to have upgraded by, probably a month, 180 epochs after `start_epoch`.
+5. `threshold` should be 90% or 75% for testnet.
+6. `minimum_activation_epoch` should be set to several epochs after `timeout_epoch` if the `start_epoch` is to be very soon after software with parameters is expected to be released.
+Where the locked_in threshold is reached, softforks are guaranteed to activate eventually but not until `minimum_activation_epoch` after signal tracking starts, allowing users, developers, and organizations to prepare software, announcements, and celebrations for that event.
+
+### States
+
+With each epoch and softfork, we associate a deployment state. The possible states are:
+
+1. DEFINED is the first state that each softfork starts. The genesis epoch is by definition in this state for each deployment.
+2. STARTED for epochs past the `start_epoch`.
+3. LOCKED_IN for one epoch after the first epoch period with STARTED epochs of which at least `threshold` has the associated bit set in `version`.
+4. ACTIVE for all epochs after the LOCKED_IN epoch.
+5. FAILED for one epoch period past the `timeout_epoch`, if LOCKED_IN was not reached.
+
+### Bit flags
+
+The `version` block header field is to be interpreted as a 32-bit little-endian integer (as present), and bits are selected within this integer as values (1 << N) where N is the bit number.
+
+Blocks in the STARTED state get a `version` whose bit position bit is set to 1. The top 3 bits of such blocks must be 001, so the range of possible `version` values is [0x20000000...0x3FFFFFFF], inclusive.
+
+By restricting the top 3 bits to 001, we get 29 out of those for this proposal and support future upgrades for different mechanisms. When a block `version` does not have top bits 001, it is treated as if all bits are 0 for deployments.
+
+Miners should continue setting the bit in the LOCKED_IN phase, so uptake is visible, though this does not affect consensus rules.
+
+### New consensus rules
+The new consensus rules for each softfork are enforced for each block with an ACTIVE state.
+
+### State transitions
+
+![State transitions](images/state-transitions.png)
+
+The genesis epoch has a state DEFINED for each deployment, by definition.
```

Comment:
It seems "genesis epoch" means the beginning epoch of a softfork proposal in this RFC? If so I think it's a bad name 'cause it's easy to mix it with the epoch 0 of CKB chain.

Reply:
No, it does means the epoch 0 of CKB chain.
This proposal allowing multiple deploy in parallel, so multiple parallel deployments will have different states, but we specify that each deployment will start from DEFINED state, with genesis as the starting point, until reach the specified start_epoch, then the state will change to started.